### PR TITLE
Refs [TASK][40625] Bind data, refresh, loadmore

### PR DESCRIPTION
**Pull request by okuloj**
## Related Tickets
- [#Task 40625: Bind data, refresh, loadmore](https://edu-redmine.sun-asterisk.vn/issues/40625)
## WHAT
- bind data, refresh, loadmore
## Evidence (Screenshot or Video)
- 

https://user-images.githubusercontent.com/55242770/130400596-8645bb1f-af5f-4719-9a27-19e8ab7f7735.mp4

## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://github.com/framgia/coding-standards/tree/master/eng/android |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/edu/news/screen/home/articles/ArticlesFragment.kt [hunk 1]
```diff
@@ -13,21 +16,45 @@ import kotlinx.android.synthetic.main.fragment_articles.view.*
 
 class ArticlesFragment : BaseFragment(), ArticlesContract.View {
 
-    private  val articlesAdapter by lazy { ArticlesAdapter() }
+    private lateinit var articlePresenter: ArticlePresenter
+    private lateinit var categoryType: CategoryType
+    private val articlesAdapter by lazy { ArticlesAdapter() }
+    private val articlesLayoutManager by lazy { LinearLayoutManager(requireContext()) }
```
Comment (luongth-2520) [inline app/src/main/java/com/edu/news/screen/home/articles/ArticlesFragment.kt]: layoutManager có thể gán trực tiếp trong xml

File: app/src/main/java/com/edu/news/screen/home/articles/ArticlesFragment.kt [hunk 2]
```diff
@@ -13,21 +16,45 @@ import kotlinx.android.synthetic.main.fragment_articles.view.*
...
+            layoutManager = articlesLayoutManager
+            addItemDecoration(dividerItemDecoration)
+        }
+        view.refreshArticles.run {
+            setOnRefreshListener {
+                articlePresenter.getArticles(categoryType)
+                isRefreshing = false
+            }
+        }
+        view.recyclerViewArticles.addOnScrollListener(object : RecyclerView.OnScrollListener() {
```
Comment (luongth-2520) [inline app/src/main/java/com/edu/news/screen/home/articles/ArticlesFragment.kt]: scope function

File: app/src/main/res/layout/fragment_home.xml [hunk 3]
```diff
@@ -46,28 +46,29 @@
...
             android:layout_height="wrap_content"
-            android:layout_marginLeft="@dimen/dp_10"
+            android:layout_marginStart="@dimen/dp_10"
             android:text="@string/category"
             android:textColor="@color/color_4A78C9"
             android:textSize="@dimen/sp_24"
             android:textStyle="bold" />
 
         <ImageView
+            android:id="@+id/imgConfigCategory"
```
Comment (luongth-2520) [inline app/src/main/res/layout/fragment_home.xml]:
```suggestion
            android:id="@+id/imageConfigCategory"
```

File: app/src/main/java/com/edu/news/data/source/remote/ArticleRemoteDataSource.kt [hunk 4]
```diff
@@ -11,11 +11,15 @@ class ArticleRemoteDataSource : ArticleDataSource.Remote {
...
-        var key = if (categoryType == CategoryType.TOP_HEADLINE) ""
-        else Constant.BASE_CATEGORY_PARAM + categoryType.nameType
+    override fun getArticles(
+        listener: OnFetchDataListener<MutableList<Article>>,
+        categoryType: CategoryType,
+        pageNumber: Int
+    ) {
+        val key = if (categoryType == CategoryType.TOP_HEADLINE) Constant.BASE_PAGE_PARAM + pageNumber
+        else Constant.BASE_CATEGORY_PARAM + categoryType.nameType + Constant.BASE_PAGE_PARAM + pageNumber
         apiService.callAPI(listener, urlGetArticles, key)
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/edu/news/data/source/remote/ArticleRemoteDataSource.kt]:
```suggestion
        val key = if (categoryType == CategoryType.TOP_HEADLINE) Constant.BASE_PAGE_PARAM
        else Constant.BASE_CATEGORY_PARAM + categoryType.nameType + Constant.BASE_PAGE_PARAM
        apiService.callAPI(listener, urlGetArticles, key + pageNumber)
```

File: app/src/main/java/com/edu/news/screen/home/articles/ArticlesAdapter.kt [hunk 5]
```diff
@@ -8,38 +8,81 @@ import com.edu.news.data.model.Article
...
-                textViewTitle.text = article.title
-                textViewDateTime.text = article.publishedAt
-                LoadImageFromUrl(imageArticle).loadImage(article.urlToImage)
+
+        fun bind(article: Article) = with(itemView) {
+            article.let {
+                textViewTitle.text = it.title
+                textViewDateTime.text =
+                    LocalDateTime.parse(it.publishedAt, DateTimeFormatter.ISO_DATE_TIME)
+                        .format(DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm"))
```
Comment (luongth-2520) [inline app/src/main/java/com/edu/news/screen/home/articles/ArticlesAdapter.kt]: đưa vào strings

File: app/src/main/java/com/edu/news/screen/home/articles/ArticlesAdapter.kt [hunk 6]
```diff
@@ -8,38 +8,81 @@ import com.edu.news.data.model.Article
...
     class ArticleViewHolder(view: View) : RecyclerView.ViewHolder(view) {
-        fun bind(article: Article) {
-            itemView.apply {
-                textViewTitle.text = article.title
-                textViewDateTime.text = article.publishedAt
-                LoadImageFromUrl(imageArticle).loadImage(article.urlToImage)
+
+        fun bind(article: Article) = with(itemView) {
+            article.let {
+                textViewTitle.text = it.title
```
Comment (luongth-2520) [inline app/src/main/java/com/edu/news/screen/home/articles/ArticlesAdapter.kt:67]: ở đây sao lại dùng let ?

File: app/src/main/res/layout/item_article.xml [hunk 7]
```diff
@@ -1,55 +1,78 @@
...
-        <ImageView
-            android:id="@+id/imageDot"
+        <androidx.constraintlayout.widget.ConstraintLayout
             android:layout_width="0dp"
-            android:layout_height="0dp"
-            android:layout_marginStart="@dimen/dp_4"
-            android:src="@drawable/ic_baseline_circle_24"
+            android:layout_height="wrap_content"
+            android:layout_margin="@dimen/dp_6"
+            android:layout_marginBottom="4dp"
```
Comment (tuanlvt-1117) [inline app/src/main/res/layout/item_article.xml]: hard code

File: app/src/main/java/com/edu/news/screen/main/MainActivity.kt [hunk 8]
```diff
@@ -20,7 +22,10 @@ class MainActivity : BaseActivity() {
             FavoriteFragment.newInstance(),
             FollowsFragment.newInstance()
         )
-        viewPagerBottomNavigation.adapter = BottomNavAdapter(this, fragments)
+        viewPagerBottomNavigation.apply {
+            isUserInputEnabled = false
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/edu/news/screen/main/MainActivity.kt:26]: hàm này dùng để làm gì e?

File: app/src/main/java/com/edu/news/screen/home/articles/ArticlesAdapter.kt [hunk 9]
```diff
@@ -8,38 +8,81 @@ import com.edu.news.data.model.Article
...
 import kotlinx.android.synthetic.main.item_article.view.*
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
 
-class ArticlesAdapter : RecyclerView.Adapter<ArticlesAdapter.ArticleViewHolder>() {
+class ArticlesAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
 
-    private val articles = mutableListOf<Article>()
+    private val articles = mutableListOf<Article?>()
+    private var pageNumber = 1
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/edu/news/screen/home/articles/ArticlesAdapter.kt]: page được xử lý tại presenter.
Khi view chạy loadmore thì sẽ gọi đến fun getArtice trong presenter và truyền vào số page tương ứng

File: app/src/main/java/com/edu/news/screen/home/articles/ArticlesContract.kt [hunk 10]
```diff
@@ -8,10 +8,12 @@ class ArticlesContract {
 
     interface View {
         fun onGetArticlesSuccess(articles: MutableList<Article>)
+        fun onLoadMoreArticlesSuccess(articles: MutableList<Article>)
         fun onError(exception: Exception?)
     }
 
     interface Presenter : BasePresenter<View> {
         fun getArticles(categoryType: CategoryType)
+        fun loadMoreArticles(categoryType: CategoryType, pageNumber: Int)
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/edu/news/screen/home/articles/ArticlesContract.kt]: 1 fun là đủ 
fun getArticles(categoryType: CategoryType, pageNumber: Int)

File: app/src/main/java/com/edu/news/screen/home/articles/ArticlesContract.kt [hunk 11]
```diff
@@ -8,10 +8,12 @@ class ArticlesContract {
 
     interface View {
         fun onGetArticlesSuccess(articles: MutableList<Article>)
+        fun onLoadMoreArticlesSuccess(articles: MutableList<Article>)
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/edu/news/screen/home/articles/ArticlesContract.kt]: 1 cái success là đủ

File: app/src/main/java/com/edu/news/screen/main/MainActivity.kt [hunk 12]
```diff
@@ -1,6 +1,8 @@
 package com.edu.news.screen.main
 
+import androidx.core.content.ContextCompat
```
Comment (duchh-2641) [inline app/src/main/java/com/edu/news/screen/main/MainActivity.kt]: ContextCompat dùng ở đâu nhỉ?

File: app/src/main/java/com/edu/news/screen/home/articles/ArticlesFragment.kt [hunk 13]
```diff
@@ -13,29 +16,62 @@ import kotlinx.android.synthetic.main.fragment_articles.view.*
 
 class ArticlesFragment : BaseFragment(), ArticlesContract.View {
 
-    private  val articlesAdapter by lazy { ArticlesAdapter() }
+    private lateinit var articlePresenter: ArticlePresenter
+    private lateinit var categoryType: CategoryType
+    private val articlesAdapter by lazy { ArticlesAdapter() }
+    private var pageNumber = 1
```
Comment (luongth-2520) [inline app/src/main/java/com/edu/news/screen/home/articles/ArticlesFragment.kt:22]: pageNumber để bên presenter

**Review by luongth-2520 — APPROVED** [review]

**Review by duchh-2641 — APPROVED** [review]

**Review by tuanlvt-1117 — APPROVED** [review]

**Pull request merged**